solana/indexer/backfill.go

```diff
@@ -0,0 +1,246 @@
+			return fmt.Errorf("no signatures found in block at slot %d", fromSlot)
+		}
+		txRange.until = block.Signatures[0]
+	}
+
+	var wg sync.WaitGroup
+	for _, address := range []solana.PublicKey{
+		s.config.SolanaConfig.RewardManagerProgramID,
+		s.config.SolanaConfig.ClaimableTokensProgramID,
+		s.config.SolanaConfig.PaymentRouterProgramID,
+	} {
+		wg.Add(1)
+		go func(address solana.PublicKey) {
+			defer wg.Done()
+			s.backfillAddressTransactions(ctx, address, txRange, fromSlot, toSlot)
+		}(address)
+	}
+	wg.Wait()
+
+	return nil
+}
+
+// Fetches and processes transactions for a given address within a given signature/slot range.
+func (s *SolanaIndexer) backfillAddressTransactions(ctx context.Context, address solana.PublicKey, txRange transactionRange, fromSlot uint64, toSlot uint64) {
+	var lastIndexedSig solana.Signature
+	foundIntersection := false
+	before := txRange.before
+
+	logger := s.logger.With(
+		zap.String("indexerSource", "rpcBackfill"),
+		zap.String("address", address.String()),
+	)
+	logger.Info("starting backfill for address",
+		zap.String("before", txRange.before.String()),
+		zap.String("until", txRange.until.String()),
+		zap.Uint64("fromSlot", fromSlot),
+		zap.Uint64("toSlot", toSlot),
+	)
+
+	limit := 1000
+	opts := rpc.GetSignaturesForAddressOpts{
+		Commitment:     rpc.CommitmentConfirmed,
+		Until:          txRange.until,
+		Limit:          &limit,
+		MinContextSlot: &toSlot,
+	}
+
+	for !foundIntersection {
+		select {
+		case <-ctx.Done():
+			logger.Error("failed to find intersection", zap.Error(ctx.Err()))
+			return
+		default:
+		}
+
+		opts.Before = before
+		res, err := withRetries(func() ([]*rpc.TransactionSignature, error) {
+			return s.rpcClient.GetSignaturesForAddressWithOpts(ctx, address, &opts)
+		}, 5, time.Second*1)
+		if err != nil {
+			logger.Error("failed to get signatures for address", zap.Error(err))
+			continue
+		}
+
+		if len(res) == 0 {
+			logger.Info("no transactions left to index.")
+			break
+		}
+
+		for _, sig := range res {
+			logger := logger.With(zap.String("signature", sig.Signature.String()))
+
+			select {
+			case <-ctx.Done():
+				logger.Error("failed to process signature", zap.Error(ctx.Err()))
+				return
+			default:
+			}
+
+			if sig.Slot < fromSlot {
+				foundIntersection = true
+				logger.Info("found intersection with transaction range")
+				break
+			}
+
+			// Skip error transactions
+			if sig.Err != nil {
+				lastIndexedSig = sig.Signature
+				continue
+			}
+
+			// Skip zero signatures
+			if sig.Signature.IsZero() {
+				continue
+			}
+
+			// Skip existing transactions
+			var exists bool
+			checkSql := `SELECT EXISTS(SELECT 1 FROM sol_token_account_balance_changes WHERE signature = @signature)`
+			err := s.pool.QueryRow(context.Background(), checkSql, pgx.NamedArgs{
+				"signature": sig.Signature,
+			}).Scan(&exists)
+
+			if err != nil {
+				logger.Error("failed to check if signature exists", zap.Error(err))
+				continue
```

Comment:
Does this have the effect of dropping the tx if the database returns an error for some reason unrelated to this? I'm still trying to wrap my head around the overall process here.

Reply:
this will actually stall - the loop will try the same tx again over and over

Reply:
which... admittedly, is probably not great...